pkg/utils/utils.go

```diff
@@ -200,8 +201,71 @@ func DiscoverSriovDevices(withUnsupported bool, storeManager StoreManagerInterfa
 }
 
 // SyncNodeState Attempt to update the node state to match the desired state
-func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool) error {
-	return ConfigSriovInterfaces(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool, parallelNicConfig bool) error {
+	if !parallelNicConfig {
+		return ConfigSriovInterfaces(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+	}
+	return ConfigSriovInterfacesInParallel(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+}
+
+func ConfigSriovInterfacesInParallel(interfaces []sriovnetworkv1.Interface, ifaceStatuses []sriovnetworkv1.InterfaceExt, pfsToConfig map[string]bool) error {
+	log.Log.V(2).Info("ConfigSriovInterfacesInParallel(): start sriov configuration")
+	if IsKernelLockdownMode(true) && hasMellanoxInterfacesInSpec(ifaceStatuses, interfaces) {
+		log.Log.Error(nil, "cannot use mellanox devices when in kernel lockdown mode")
+		return fmt.Errorf("cannot use mellanox devices when in kernel lockdown mode")
+	}
+	// TODO(e0ne): store all errors in SriovNetworkNodeState
+	var result error
+	wg := sync.WaitGroup{}
+	for _, ifaceStatus := range ifaceStatuses {
```

Comment:
I believe the concurrency should start here, for each one of the "ifaceStatuses". This way, you'll also cover the resetting of the sriov as @adrianchiris mentioned. 

There is also a well-known pattern in concurrency Go for this type of operations. Basically, the routines send the result through a channel and there is a for loop waiting and aggregating the results. I'll paste a working example here where I also address the defer and not reference for the loop var comments of @zeeke. 

func ConfigSriovInterfaces(interfaces []sriovnetworkv1.Interface, ifaceStatuses []sriovnetworkv1.InterfaceExt, pfsToConfig map[string]bool) error {
	if IsKernelLockdownMode(true) && hasMellanoxInterfacesInSpec(ifaceStatuses, interfaces) {
		log.Log.Error(nil, "cannot use mellanox devices when in kernel lockdown mode")
		return fmt.Errorf("cannot use mellanox devices when in kernel lockdown mode")
	}

	// we are already inside chroot, so we initialize the store as running on host
	storeManager, err := NewStoreManager(true)
	if err != nil {
		return fmt.Errorf("SyncNodeState(): error initializing storeManager: %v", err)
	}

	var wg sync.WaitGroup
	resChannel := make(chan error)
	var result []error

	for _, ifaceStatus := range ifaceStatuses {
		wg.Add(1)
		go func(ifaceStatus sriovnetworkv1.InterfaceExt) {
			var err error
			defer func() {
				resChannel <- err
				wg.Done()
			}()

			configured := false
			for _, iface := range interfaces {
				if iface.PciAddress == ifaceStatus.PciAddress {
					configured = true

					if skip := pfsToConfig[iface.PciAddress]; skip {
						break
					}

					if !NeedUpdate(&iface, &ifaceStatus) {
						log.Log.V(2).Info("syncNodeState(): no need update interface", "address", iface.PciAddress)

						// Save the PF status to the host
						err = storeManager.SaveLastPfAppliedStatus(&iface)
						if err != nil {
							log.Log.Error(err, "SyncNodeState(): failed to save PF applied config to host")
							return
						}

						break
					}
					if err = configSriovDevice(&iface, &ifaceStatus); err != nil {
						log.Log.Error(err, "SyncNodeState(): fail to configure sriov interface. resetting interface.", "address", iface.PciAddress)
						if iface.ExternallyManaged {
							log.Log.Info("SyncNodeState(): skipping device reset as the nic is marked as externally created")
						} else {
							if resetErr := resetSriovDevice(ifaceStatus); resetErr != nil {
								log.Log.Error(resetErr, "SyncNodeState(): failed to reset on error SR-IOV interface")
							}
						}
						return
					}

					// Save the PF status to the host
					err = storeManager.SaveLastPfAppliedStatus(&iface)
					if err != nil {
						log.Log.Error(err, "SyncNodeState(): failed to save PF applied config to host")
						return
					}
					break
				}
			}
			if !configured && ifaceStatus.NumVfs > 0 {
				if skip := pfsToConfig[ifaceStatus.PciAddress]; skip {
					return
				}

				// load the PF info
				var pfStatus *sriovnetworkv1.Interface
				var exist bool
				pfStatus, exist, err = storeManager.LoadPfsStatus(ifaceStatus.PciAddress)
				if err != nil {
					log.Log.Error(err, "SyncNodeState(): failed to load info about PF status for device",
						"address", ifaceStatus.PciAddress)
					return
				}

				if !exist {
					log.Log.Info("SyncNodeState(): PF name with pci address has VFs configured but they weren't created by the sriov operator. Skipping the device reset",
						"pf-name", ifaceStatus.Name,
						"address", ifaceStatus.PciAddress)
					return
				}

				if pfStatus.ExternallyManaged {
					log.Log.Info("SyncNodeState(): PF name with pci address was externally created skipping the device reset",
						"pf-name", ifaceStatus.Name,
						"address", ifaceStatus.PciAddress)
					return
				} else {
					err = RemoveUdevRule(ifaceStatus.PciAddress)
					if err != nil {
						return
					}
				}

				if err = resetSriovDevice(ifaceStatus); err != nil {
					return
				}
			}
		}(ifaceStatus)
	}

	// Wait for routines to finish and close resChannel to exit below for loop.
	// Routines only decrement counter when error is read in the for loop.
	go func() {
		wg.Wait()
		close(resChannel)
	}()

	// Fetch error returned by the routines. This will exit after channel is closed. No more pending routines.
	for routineErr := range resChannel {
		if routineErr != nil {
			result = append(result, routineErr)
		}
	}

	// Only first arriving error is returned.
	if len(result) != 0 {
		log.Log.Error(result[0], "ConfigSriovInterfacesInParallel(): fail to configure sriov interfaces. Only first arriving error is returned. There might be more. Please check the logs.")
		return result[0]
	}

	return nil
}

Note that I'm returning the first error of the result list of errors. For other projects, I've printed and returned this whole list. But printing the errors is already done separately and for returning a list, we will need other changes.

For the storage manager, I used a mutex and lock unlock in save and load.

type StoreManager struct {
	RunOnHost bool
	sync.Mutex
}

func (s *StoreManager) SaveLastPfAppliedStatus(PfInfo *sriovnetworkv1.Interface) error {
	s.Lock()
	defer s.Unlock()
}

Also note that everything after "configured := false" can be encapsulated in a function. So we could avoid duplication.

I've tested this but not extensively. Just wanted to show you the pattern I've used in other projects.